Add Environment Min Light property to LightmapGI

Calinou opened the pull request:
This property works the same as BakedLightmap's Environment Min Light property in `3.x`. It can be used to avoid overly dark areas in the lightmap, or to tweak the shadow color for artistic reasons.

This closes https://github.com/godotengine/godot/issues/50568.

**Testing project:** [test-lightmapgi-environment-min-light.zip](https://github.com/user-attachments/files/19398744/test-lightmapgi-environment-min-light.zip)

## Preview

*The color is exaggerated for previewing purposes. Typically, you'll use much darker colors like `#050505`.*

![2021-07-18_00 50 11](https://user-images.githubusercontent.com/180032/126051205-671055f4-c7b3-46fc-8fab-e2ac13f8746f.png)

## TODO

- [ ] Make lightmap probes affected by Environment Min Light so that dynamic objects are lit correctly.
- [ ] Fix Environment Min Light being too bright when Directional is enabled.
  - No changes would be needed if https://github.com/godotengine/godot/pull/100790 is merged, as it'll store directional data in a separate texture.

Calinou committed:
Add Environment Min Light property to LightmapGI

This property works the same as BakedLightmap's Environment Min Light
property in `3.x`. It can be used to avoid overly dark areas in the
lightmap, or to tweak the shadow color for artistic reasons.

reduz commented:
I do not see the point of this given you have the ambient constant color, its quite redundant.

Calinou commented:
> I do not see the point of this given you have the ambient constant color, its quite redundant.

The constant ambient color does not spread to shaded areas when using LightmapGI. While this is physically accurate, it can result in scenes looking too dark unless you manually add a bunch of OmniLight3Ds in dark spots (unmotivated light sources).

ScarfKat commented:
This would be super helpful. I just ran into a relevant issue today, with lightmapping being darker than I anticipated. Fingers crossed this makes it into 4.4.

Calinou commented:
Rebased and tested again, it works as expected.

> This would be super helpful. I just ran into a relevant issue today, with lightmapping being darker than I anticipated. Fingers crossed this makes it into 4.4.

I'm afraid there are several outstanding issues I need to address before this is in a mergeable state (see TODO in the pull request description). These are nontrivial to fix as I'll need to modify the lightmap baking shader to bake the correct data on the GPU, instead of modifying the final image data on the CPU (which can't affect lightmap probes, or deal with directional lightmaps correctly). This is more involved since I'm not familiar with the lightmap baking shader's code.

Calinou commented:
Rebased and tested again, it works as expected.

I've added code to modify probe colors according to Environment Min Light and fixed the colorspace used to modify the lightmap (the color is now converted from sRGB to linear before being applied).

However, the probe color is not uniformly applied and it introduces an "inverse color" effect in the direction opposite of the incoming light:

![image](https://github.com/user-attachments/assets/73da4213-e2cf-4402-92cf-3ad35916ab73)

This also applies if using a grayscale color. In this case, the light on dynamic objects will be more contrasted compared to not using any Environment Min Light color.

ScarfKat commented:
Still hoping for this! I'd be glad to use baked lighting in my projects, but I tend to always find the dark sections of it to be too dark. This would make the feature considerably more useful across the board.
